pkg/v1/tkg/client/package_helper.go

```diff
@@ -0,0 +1,102 @@
+		name := strings.TrimSuffix(clusterBootstrap.Spec.Kapp.ValuesFrom.ProviderRef.Name, suffixStr)
+		packages = append(packages, kapppkgv1alpha1.Package{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: clusterBootstrap.ObjectMeta.Namespace}})
+	}
+	if clusterBootstrap.Spec.CNI != nil && clusterBootstrap.Spec.CNI.ValuesFrom != nil && clusterBootstrap.Spec.CNI.ValuesFrom.ProviderRef != nil {
+		name := strings.TrimSuffix(clusterBootstrap.Spec.CNI.ValuesFrom.ProviderRef.Name, suffixStr)
+		packages = append(packages, kapppkgv1alpha1.Package{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: constants.CorePackagesNamespace}})
+	}
+	if clusterBootstrap.Spec.CSI != nil && clusterBootstrap.Spec.CSI.ValuesFrom != nil && clusterBootstrap.Spec.CSI.ValuesFrom.ProviderRef != nil {
+		name := strings.TrimSuffix(clusterBootstrap.Spec.CSI.ValuesFrom.ProviderRef.Name, suffixStr)
+		packages = append(packages, kapppkgv1alpha1.Package{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: constants.CorePackagesNamespace}})
+	}
+	if clusterBootstrap.Spec.CPI != nil && clusterBootstrap.Spec.CPI.ValuesFrom != nil && clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef != nil {
+		name := strings.TrimSuffix(clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef.Name, suffixStr)
+		packages = append(packages, kapppkgv1alpha1.Package{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: constants.CorePackagesNamespace}})
+	}
+	return packages
+}
+
+// MonitorAddonsCorePackageInstallation monitors addon's core packages (kapp, cni, csi and cpi) and returns error if any while monitoring packages or any packages are not installed successfully. First it monitors kapp package in management cluster then it monitors other core packages in workload cluster.
+func MonitorAddonsCorePackageInstallation(regionalClusterClient clusterclient.Client, workloadClusterClient clusterclient.Client, packages []kapppkgv1alpha1.Package, packageInstallTimeout time.Duration) error {
+	if len(packages) == 0 {
+		return nil
+	}
+	var kappPackage, corePackages []string
+	var kappNamespace string
+	for _, p := range packages {
+		if strings.Contains(p.ObjectMeta.Name, "kapp-controller") {
+			kappPackage = append(kappPackage, p.ObjectMeta.Name)
+			kappNamespace = p.ObjectMeta.Namespace
+		} else {
+			corePackages = append(corePackages, p.ObjectMeta.Name)
+		}
+	}
+	log.V(3).Infof("list of addons core packages from ClusterBootstrap object: %v", append(kappPackage, corePackages...))
+	err := WaitForPackagesInstallation(regionalClusterClient, kappPackage, kappNamespace, packageInstallTimeout)
+	if err != nil {
+		return err
+	}
+	return WaitForPackagesInstallation(workloadClusterClient, corePackages, constants.CorePackagesNamespace, packageInstallTimeout)
+}
+
+func WaitForPackagesInstallation(clusterClient clusterclient.Client, packageInstallNames []string, namespace string, packageInstallTimeout time.Duration) error {
```

Comment:
Why should this function be exported?

Reply:
It's an independent function, other packages can use it if needed.